Add UOps.VECTORIZE [run_process_replay]

greg-niemeyer opened the pull request.

greg-niemeyer commented:
This change was made in response to the [tinygrad-dev](https://discord.com/channels/1068976834382925865/1068982781490757652/1257903549355135089) conversation between geohot & qazalin:

> geohot: UOps.CAST currently means two things, a typechange and putting types together. someone want to add a new uop and fix this? open to name suggestions
> 
> qazalin: I'd pick UOps.VECTORIZE, keeping with the naming in dtypes.vec
> 
> geohot: What’s the LLVMs name for this?
> 
> geohot: oh, they store in the gep it looks like kk, VECTORIZE it is.
> Someone here want to earn a purple and do it?

Qazalin commented on tinygrad/codegen/uops.py:
```diff
@@ -298,18 +299,18 @@ def loop_collapse(loop_start, loop_end, compval, idx, mval, multconst, rng):
-  (UPat(UOps.CAST, name="root", src=tuple(UPat(UOps.PHI, src=(UPat(UOps.GEP, i, src=(UPat(name="val"),)), UPat(name=f"v{i}"))) for i in range(2))),
-    lambda root, val, v0, v1: UOp(UOps.PHI, root.dtype, (val, UOp(UOps.CAST, val.dtype, (v0, v1))))),
+  # store float4/float2 directly (remove VECTORIZE/GEP)
+  (UOp.store(UOp.var("buf"), UOp.var("idx"), UOp(UOps.VECTORIZE, src=tuple(UOp(UOps.GEP, arg=i, src=(UOp.var("val"),)) for i in range(4)))), UOp.store), # noqa: E501
```
We'd prefer to break into multiple lines if it hits the line length limit.

Qazalin commented on tinygrad/renderer/cstyle.py:
```diff
@@ -146,7 +146,7 @@ def ssa(prefix:str, u:Optional[UOp]=None):
-        elif uop in {UOps.CAST, UOps.BITCAST}:
+        elif uop in {UOps.CAST, UOps.BITCAST, UOps.VECTORIZE}:
           if uop is UOps.BITCAST:
             assert len(src) == 1
```
this assert should also work with UOps.CAST.

Qazalin commented on tinygrad/renderer/llvmir.py:
```diff
@@ -146,7 +146,7 @@ def render(self, name:str, uops:UOpGraph) -> str:
         elif uop is UOps.ALU:
           lvars[u] = code_for_op[args](bb[-1], *[lvars[x] for x in src], dtype if args not in (BinaryOps.CMPLT, BinaryOps.CMPNE) else src[0].dtype)
-        elif uop in {UOps.CAST, UOps.BITCAST}: lvars[u] = cast(bb, lvars[src[0]], src[0].dtype, dtype, bitcast=uop is UOps.BITCAST)
+        elif uop in {UOps.CAST, UOps.BITCAST, UOps.VECTORIZE}: lvars[u] = cast(bb, lvars[src[0]], src[0].dtype, dtype, bitcast=uop is UOps.BITCAST)
```
llvm doesn't have vectorize.

Qazalin commented on tinygrad/codegen/uops.py:
```diff
@@ -298,18 +299,18 @@ def loop_collapse(loop_start, loop_end, compval, idx, mval, multconst, rng):
   (UOp.lt(-UOp.var('x'), UOp.cvar('c', dtypes.int)), lambda c,x: UOp.lt(UOp.const(c.dtype, -c.arg), x)),
   # cast NOOP (NOTE: it's str to deal with PtrDType)
-  (UPat(UOps.CAST, name="root"), lambda root: root.src[0] if str(root.dtype) == str(root.src[0].dtype) else None),
+  (UPat(UOps.VECTORIZE, name="root"), lambda root: root.src[0] if str(root.dtype) == str(root.src[0].dtype) else None),
```
hmm this is wrong actually, the CAST rule should still be here.

greg-niemeyer replied:
Thanks for the review Qazalin! I wasn't able to find time to address these comments today, but I'll be on it tomorrow.

Regarding this particular line, I made this change because the update I made to [test_uop_graph.py::test_cast_vectorized_fold](https://github.com/tinygrad/tinygrad/pull/5289/files#diff-9efe3eddc1a3cba3b388b938df1a00c456d555f054802b9edd448e753c92e3f7L123-L135) failed without it. I'll investigate further myself, but do you know if this test should be exercising UOps.Vectorize, or should I revert the change I made to that test along with this line?

Qazalin replied:
you need both, added a test for CAST here https://github.com/tinygrad/tinygrad/pull/5292 - you can see it in the rebase.

Qazalin commented:
diff looks clean, last merge blocker is these tensor cores having an extra UOp:

https://github.com/tinygrad/tinygrad/actions/runs/9820719351/job/27115613583?pr=5289#step:24:42

greg-niemeyer commented:
> diff looks clean, last merge blocker is these tensor cores having an extra UOp:
> 
> https://github.com/tinygrad/tinygrad/actions/runs/9820719351/job/27115613583?pr=5289#step:24:42

Fixed and added a test to catch the missing rule 😃

Qazalin commented on test/test_uop_graph.py:
```diff
@@ -119,16 +119,25 @@ def test_const_cast(self):
 
+  def test_const_vectorize(self):
+    bf = UOp(UOps.CONST, dtypes.bool, arg=False)
+    out = UOp(UOps.VECTORIZE, dtypes.int, (bf,))
```
yea this test is actually wrong, you can't vectorize something to another dtype. you first need to cast it.
eg: bool -> CAST int -> vectorize int4
also how doesn't the uops type_verify catch scalar dtypes.int on vectorize?

greg-niemeyer replied:
Ah, yeah, I've just been thinking about vectorize as a variant of cast, but what you're saying makes sense. I'll take another stab at it and think a little harder about what vectorize should be doing.

greg-niemeyer replied:
> also how doesn't the uops type_verify catch scalar dtypes.int on vectorize?
I think this is because the type_verify happens after the fold, so the VECTORIZE is removed before the check happens. But if I take out the rule that I added, it does indeed hit the assert.

Qazalin replied:
ahh right - It's ok if it's just fixed for this test.

greg-niemeyer replied:
> good moment to also add the dtype assert.

I added an assert, lmk what you think: https://github.com/tinygrad/tinygrad/pull/5289/files#diff-91ca5e2e75ef3ea1982c8ca6cc175ee88f20efa0d8e4b96f305b970dc6df71e7R102

Trying to assert "dtype should be a vec()'d version of src.dtype"

greg-niemeyer replied:
Oh and I think I made my test sane. I constructed it based on what I was seeing in the prior process_replay failures as opposed to doing nonsensical things with the dtypes lol.

Qazalin commented on tinygrad/renderer/cstyle.py:
```diff
@@ -58,9 +62,9 @@ def render_load(self, output_dtype, buf_name, buf_dtype, idx, local=False) -> st
-    else:
-      out_val = f"*({buf_name}+{idx})" if self.uses_ptr_arithmetic else f"{buf_name}[{idx}]"
-    return self.render_cast([out_val], output_dtype) if output_dtype != buf_dtype else out_val
+      return self.render_cast(out_val, output_dtype) if output_dtype != buf_dtype else out_val
```
hmm is this diff required?

greg-niemeyer replied:
pylint was complaining about this in my pre-commit validation (I assume because I touched the line removing brackets from out_val). I can revert this change and push just my edit if desired.

Qazalin replied:
yea it's better if the diff is as small as possible.

greg-niemeyer replied:
reverted the unnecessary stuff
